File: Lab 7 (14-09-23)/pc.java
class buffer{
	public int info;
	volatile int flag=0;
	buffer(){flag=0;}
	public synchronized  put(int n){
		if(flag==1) {
		try{ 
			wait();
		 } 
		catch(InterruptedException e){
			System.out.println("Interrupted Exception");
		}
		}
		flag=1; 
		info=n;
		notify();
	}
	public synchronized get(){
		if(flag==0) {
		try{ 
			wait();
		 } 
		catch(InterruptedException e){
			System.out.println("Interrupted Exception");
		}
		}
		System.out.println("Information : " + info);
		flag=0;
		notify();
	}

}
class producer : Thread{
	buffer b_ref;
	int pid;
	public producer(buffer b, int id) : base(){
		
		b_ref=b;
		pid=id;
		this.start();
	}
	public public void run(){
		for(int i=0;i<5;i++){
			b_ref.put(i); 
			try{
				Thread.sleep(100);
			} catch(InterruptedException e){
				System.out.println("Interrupted Exception");
			}
		}
	}
}
class consumer : Thread{
	buffer b_ref;
	int cid;
	consumer(buffer b, int id): base(){
		b_ref=b; 
		cid=id;
		start();
	}
	public void run(){
		for(int i=0;i<5;i++){
			b_ref.get();
			try{
				Thread.sleep(100);
			} catch(InterruptedException e){
				System.out.println("Interrupted Exception");
			}
		}
	}

}
class pc{
 public static void main(String args[]){
	buffer b_ref_main = new buffer();
	producer p_ref = new producer(b_ref_main,0);
	consumer c_ref = new consumer(b_ref_main,0);
	try
	{
		p_ref.join();
  		c_ref.join();
	}
	catch(InterruptedException e){
		System.out.println("Interrupted Exception");
	}
	
 }
}

File: MultiThreading/PC.cs
using System;
using System.Threading;
class Buffer
{
    private int info;
    private volatile int flag = 0;
    public Buffer()
    {
        flag = 0;
    }
    public void Put(int n)
    {
        lock (this)
        {
            if (flag == 1)
            {
                try
                {
                    Monitor.Wait(this);
                }
                catch (ThreadInterruptedException)
                {
                    Console.WriteLine("Interrupted Exception");
                }
            }
            flag = 1;
            info = n;
            Monitor.Pulse(this);
        }
    }
    public void Get()
    {
        lock (this)
        {
            if (flag == 0)
            {
                try
                {
                    Monitor.Wait(this);
                }
                catch (ThreadInterruptedException)
                {
                    Console.WriteLine("Interrupted Exception");
                }
            }
            Console.WriteLine("Information : " + info);
            flag = 0;
            Monitor.Pulse(this);
        }
    }
}
class Producer
{
    private Buffer b_ref;
    private int pid;
    public Producer(Buffer b, int id)
    {
        b_ref = b;
        pid = id;
        Thread t = new Thread(Run);
        t.Start();
    }
    public void Run()
    {
        for (int i = 0; i < 5; i++)
        {
            b_ref.Put(i);
            try
            {
                Thread.Sleep(100);
            }
            catch (ThreadInterruptedException)
            {
                Console.WriteLine("Interrupted Exception");
            }
        }
    }
}
class Consumer
{
    private Buffer b_ref;
    private int cid;
    public Consumer(Buffer b, int id)
    {
        b_ref = b;
        cid = id;
        Thread t = new Thread(Run);
        t.Start();
    }
    public void Run()
    {
        for (int i = 0; i < 5; i++)
        {
            b_ref.Get();
            try
            {
                Thread.Sleep(100);
            }
            catch (ThreadInterruptedException)
            {
                Console.WriteLine("Interrupted Exception");
            }
        }
    }
}
class PC
{
    public static void Main(string[] args)
    {
        Buffer b_ref_main = new Buffer();
        Producer p_ref = new Producer(b_ref_main, 0);
        Consumer c_ref = new Consumer(b_ref_main, 0);
        try
        {
            p_ref.Run();
            c_ref.Run();
        }
        catch (ThreadInterruptedException)
        {
            Console.WriteLine("Interrupted Exception");
        }
    }
}